### DISCUSSION TITLE
Implementing MAX30100

### DISCUSSION THREAD

**Discussion by William04A** [micropython/micropython#11902] [Hardware & Peripherals]
Hello!

I'm trying to make a project that measures SPO2 and heart rate in BPM using a microcontroller running MicroPython to read values. However, I'm looking into the practical implementation of it. I found a [driver](https://github.com/mfitzp/max30100/blob/master/max30100.py) and [a lot of useful information](https://morf.lv/implementing-pulse-oximeter-using-max30100). However, I'm running into a more practical problem and I'm trying to figure out if it can be solved.

The MAX30100 has a sample rate of 50-1000 samples per second, which I assume is too fast for my code. I did some measurements and figured out that it barely is able to cooperate with me. Simply reading data from I2C and appending it to a list for example took 11568us (during one measurement, others are similar). This gives me, if my math is correct and not tired from many hours coding today, the possibility to process about 80 samples per second. However, I am not getting any reasonable values with a sample rate of 50. So what I am looking for is two things:

* Am I correct in my assumption that Micropython probably is too slow to process this task and instead should use an external microcontroller?
* Does anyone happen to have a better library or more complete BPM/SPO2 calculation example lying around?

Bonus question:
This [useful website](https://morf.lv/implementing-pulse-oximeter-using-max30100) uses a Butterworth filter to filter out values.  I have no experience with this filter nor the (seemingly complicated) math behind it. If I want to recalculate the Butterworth constants for a different sampling rate, in this case, 50 hertz, can anyone point me to good references, resources, or websites that help me calculate it? The linked website in the tutorial I just linked seems useful, but to someone who is new to filters, I am just trying to figure out what to put into the boxes...

Thank you for any help!

**Comment by 2dof** [comment]
``` Am I correct in my assumption that Micropython probably is too slow to process this task and instead should use an external microcontroller? ```  
How did You design  data acquisition and processing? : sample - by - sample  or by batch?  0 I assume You want process raw data.

 - Do not append data to the list (it is slow).  -> Create array or best, ring buffer  or  solution  given by  @peterhinch [ThreadSafeQueue](https://github.com/orgs/micropython/discussions/11610).  
 - I guest You are Using ```self.buffer_red/self.buffer_ir ``` of ```class MAX30100``` to it is better to modify and use solution with buffer

```Does anyone happen to have a better library or more complete BPM/SPO2 calculation example lying around?```  based on linked MAX30100 driver by default it can sampling with 100 Hz so at least  sampling (i2c) can be handled with that speed but even sampling with 50- 60 Hz  is ok for calculating BPM and SPo2 ( effective band width 30 Hz  ( Butterworth filter cut off ) ) 

for design filter: 
[Butterworth filter design ](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.butter.html)
( In my opinion, if it possible it is always better rescale coefficients to  fixed point values ( avoiding floating point calculation) - it will speed up sample processing - unless mcu has fast floating point calculation) 

read also:
[fda-guidance-documents/pulse-oximeters](https://www.fda.gov/regulatory-information/search-fda-guidance-documents/pulse-oximeters-premarket-notification-submissions-510ks-guidance-industry-and-food-and-drug)

**Comment by peterhinch** [comment]
You might be interested in [micropython-filters](https://github.com/peterhinch/micropython-filters) which provides a means of implementing realtime digital filters on STM and RP2 chips. It would be possible to approximate a Butterworth response.
